src/jsep.js

```diff
@@ -2,257 +2,524 @@
+		this.unary_ops[op_name] = 1;
+		return this;
+	}
+
+	/**
+	 * @method jsep.addBinaryOp
+	 * @param {string} op_name The name of the binary op to add
+	 * @param {number} precedence The precedence of the binary op (can be a float)
+	 * @return this
+	 */
+	addBinaryOp(op_name, precedence) {
+		this.max_binop_len = Math.max(op_name.length, this.max_binop_len);
+		this.binary_ops[op_name] = precedence;
+		return this;
+	}
+
+	/**
+	 * @method addIdentifierChar
+	 * @param {string} char The additional character to treat as a valid part of an identifier
+	 * @return this
+	 */
+	addIdentifierChar(char) {
+		this.additional_identifier_chars.add(char);
+		return this;
+	}
+
+	/**
+	 * @method addLiteral
+	 * @param {string} literal_name The name of the literal to add
+	 * @param {*} literal_value The value of the literal
+	 * @return this
+	 */
+	addLiteral(literal_name, literal_value) {
+		this.literals[literal_name] = literal_value;
+		return this;
+	}
+
+	/**
+	 * @method removeUnaryOp
+	 * @param {string} op_name The name of the unary op to remove
+	 * @return this
+	 */
+	removeUnaryOp(op_name) {
+		delete this.unary_ops[op_name];
+		if (op_name.length === this.max_unop_len) {
+			this.max_unop_len = JsepConfig.getMaxKeyLen(this.unary_ops);
+		}
+		return this;
+	}
+
+	/**
+	 * @method removeAllUnaryOps
+	 * @return this
+	 */
+	removeAllUnaryOps() {
+		this.unary_ops = {};
+		this.max_unop_len = 0;
+
+		return this;
+	}
+
+	/**
+	 * @method removeIdentifierChar
+	 * @param {string} char The additional character to stop treating as a valid part of an identifier
+	 * @return this
+	 */
+	removeIdentifierChar(char) {
+		this.additional_identifier_chars.delete(char);
+		return this;
+	}
+
+	/**
+	 * @method removeBinaryOp
+	 * @param {string} op_name The name of the binary op to remove
+	 * @return this
+	 */
+	removeBinaryOp(op_name) {
+		delete this.binary_ops[op_name];
+
+		if (op_name.length === this.max_binop_len) {
+			this.max_binop_len = JsepConfig.getMaxKeyLen(this.binary_ops);
+		}
+
+		return this;
+	}
+
+	/**
+	 * @method removeAllBinaryOps
+	 * @return this
+	 */
+	removeAllBinaryOps() {
+		this.binary_ops = {};
+		this.max_binop_len = 0;
+
+		return this;
+	}
+
+	/**
+	 * @method removeLiteral
+	 * @param {string} literal_name The name of the literal to remove
+	 * @return this
+	 */
+	removeLiteral(literal_name) {
+		delete this.literals[literal_name];
+		return this;
+	}
+
+	/**
+	 * @method removeAllLiterals
+	 * @return this
+	 */
+	removeAllLiterals() {
+		this.literals = {};
+
+		return this;
+	}
+}
+
+
+// JavaScript Expression Parser
+export class Jsep {
+	static get version() {
+		// To be filled in by the template
+		return '<%= version %>';
+	}
+
+	/**
+	 * @returns {(pos: number) => string}
+	 */
+	get charAtFunc() {
+		return this.expr.charAt;
+	}
+
+	/**
+	 * @returns {(index: number) => number}
+	 */
+	get charCodeAtFunc() {
+		return this.expr.charCodeAt;
+	}
+
+	get _exprI() {
```

Comment:
Do we really need these? All they seem to be abstracting away is calling `charCodeAt()` and `charAt()` on `this.index`. Is this something worth abstracting? Is this something worth exposing on our public API (which means it's harder to get rid of down the line)?